client/network/src/protocol/generic_proto/upgrade/notifications.rs

```diff
@@ -169,6 +169,48 @@ where TSubstream: AsyncRead + AsyncWrite,
 
 		self.handshake = NotificationsInSubstreamHandshake::PendingSend(message.into());
 	}
+
+	/// Equivalent to `Stream::poll_next`, except that it only drives the handshake and is
+	/// guaranteed to not generate any notification.
+	pub fn poll_process(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Result<Infallible, io::Error>> {
```

Comment:
Instead of introducing another method on `NotificationsInSubstream`, why not introduce a new structure e.g. `HandshakingNotificationsInSubstream` that implements `Future` and eventually resolves in a `NotificationsInSubstream`?

I would see the following benefits:

- At compile time enforcement that once `Stream::poll_next` is called one does not go back to calling `poll_process`.

- No duplication of handshake sending code.

Reply:
I went this path because of `NotifsInHandler::poll_process`.
As far as I can tell, there isn't really any satisfying alternative way to `NotifsInHandler::poll_process`.

Taking this constraint into account, adding `NotificationsInSubstream::poll_process` was the most straight-forward way to implement `NotifsInHandler::poll_process`. Also, to be honest, a solution that is straight-forward to implement is more appealing on a Friday evening.

While we could indeed ensure at compile-time that the handshake is properly performed, we can't do the same for `NotifsInHandler`, and the runtime checks would then have to be moved to `NotifsInHandler`.